Dnd

Mxchaeltrxn opened the pull request:
Addresses #6 using https://github.com/atlassian/react-beautiful-dnd/blob/master/docs/patterns/multi-drag.md

I didn't implement all their selection methods because I think it is fine not to use hotkeys to select layers. It also may be useful to be able to use the arrow keys to move the elements around after selecting it. I did implement the drag and drop like it was described in the react-beautiful-dnd docs though.

I also think that for the feature where holding shift allows you to multi-select layers, command/control would be better because I think that's how it usually is and shift is used to select all layers inbetween the first layer clicked and the layer that was just clicked.

Thoughts?

haydn commented:
@Mxchaeltrxn Now the we've got #63 merged, do could you get this up-to-date with master? Looks like there's going to be conflicts in `Editor.js`.

> I didn't implement all their selection methods because I think it is fine not to use hotkeys to select layers.

Totally agree.

> I also think that for the feature where holding shift allows you to multi-select layers, command/control would be better because I think that's how it usually is and shift is used to select all layers inbetween the first layer clicked and the layer that was just clicked.

:+1: I've been meaning to make an issue for this.

haydn commented on src/Editor.js:
```diff
@@ -482,43 +484,130 @@ const Editor = () => {
+
+            const layers = [...doc.layers];
+            const draggedLayer = layers.splice(source.index, 1)[0];
+            layers.splice(destination.index, 0, draggedLayer);
```
We've already got a handy little function for this!

https://github.com/haydn/sketchbook-js/blob/master/src/reorder.js

haydn replied:
Might need to be modified to handle reordering multiple items at once? Probably worthwhile putting some unit tests around it to cover it current behaviour and then take a TDD approach to make it support multiple layers at a time.

Mxchaeltrxn commented:
@haydn So I haven't addressed your comment yet because I just noticed it. I can do it in a moment. 

But if you decide to look at this now, dragging the layers causes the list items to shift and because the layers are separated by a border, it looks really ugly. I've put a border around each list item to make this really obvious. I'm not entirely sure how to style it so that it looks good. I've considered removing those borders while an item is being dragged and using the div inside to create the border rather than the one outside. In my head, these are still really ugly. Thoughts?

haydn commented:
@Mxchaeltrxn Yeah, let's get this one merged next! I think there's going to be some tricky merging ahead of you, so let's get through that first and make sure everything works and we'll clean-up styles at the end.

A little bug to note in the current version: if you select a layer and then drag a different layer the new layer you dragged won't get selected and the layer that is selected will be the one that gets reordered. Might just be a visual bug.

Mxchaeltrxn commented:
@haydn Addressed that problem. I had the logic wrong. 

Notes/todos: 
1. I'm not really sure about the tests I wrote that target reordering multiple elements.
2. Style the layer elements.
3. What do you think about these lines in reorder.js? I'm trying to validate the predicates that are used as arguments. I'm asking because I'm not exactly sure how I should be testing the predicate (and therefore validating it as well).
```
  // If filtered arrays have overlap, predicates are invalid.
  if (
    elementsToShiftList.filter(elem =>
      set(multipleElementReorderedList).has(elem)
    ).length > 0
  ) {
    return "invalid predicate(s)";
  }
```

haydn commented:
@Mxchaeltrxn Alright, nice work dealing with the merge conflicts! I assume it was a bit painful.

I reckon there's probably a better interface for the `reorder` function that doesn't use predicates. Unlike the functions in `layers.js` we won't get much benefit from an [inversion of control](https://en.wikipedia.org/wiki/Inversion_of_control) pattern because we're already just passing in indices and doesn't have any special knowledge of what is in the array it's reordering — just that it's an array. We could use it for reordering layers, an array of strings (as we do in the tests) or anything else we want to. That's the same goal we're trying to achieve by using predicates for the functions in `layers.js` — we're trying to minimise the knowledge those functions have of the data they're operating on (they only know this items passed to them have `x1`, `y1`, `x2` and `y2` values, but nothing else).

With that in mind, I reckon we can make is so the `startIndex` argument can be either a single index or an array of indices (these test are just a handful of examples, it's still good to include all those edges case you've covered in your tests):

```js
test("single element, move to end", () => {
  expect(reorder(["a", "b", "c", "d"], 0, 3)).toEqual(["b", "c", "d", "a"]);
});
test("multiple elements, move to start", () => {
  expect(reorder(["a", "b", "c", "d"], [0, 3], 0)).toEqual([
    "a",
    "d",
    "b",
    "c"
  ]);
});
test("multiple elements, move to middle", () => {
  expect(reorder(["a", "b", "c", "d"], [0, 3], 1)).toEqual([
    "b",
    "a",
    "d",
    "c"
  ]);
});
test("multiple elements, move to end", () => {
  expect(reorder(["a", "b", "c", "d"], [0, 3], 2)).toEqual([
    "b",
    "c",
    "a",
    "d"
  ]);
});
test("ignore duplicates", () => {
  expect(reorder(["a", "b", "c", "d"], [0, 0], 3)).toEqual([
    "b",
    "c",
    "d",
    "a"
  ]);
});
```

I'm assuming we'll be able to use this something along these lines, however, this might might need adjustment because I'm not 100% sure what the value of `destination.index` is:

```js
const onDragEnd = ({ destination }) => {
  if (!destination) return;
  setState(current => ({
    ...current,
    doc: {
      ...current.doc,
      layers: reorder(
        current.doc.layers,
        current.doc.layers
          .map(({ id }, index) => ({ id, index }))
          .filter(({ id }) => selection.has(id))
          .map(({ index }) => index),
        destination.index
      )
    }
  }));
};
```

Mxchaeltrxn commented:
@haydn Yeah it was a little annoying to merge. Hmm I didn't realise that inversion of control wasn't that useful here. It did make things much easier to do though. 

So I've fixed the function and added relevant tests. The way you setState there is super cool. I'm surprised it took so few lines of code to do achieve that result. I've changed it a little to put the source.index at the start of the array so it's compatible with the reorder function.

haydn commented:
@Mxchaeltrxn Sorry for the delay getting back to you!

Let's use [ESDoc syntax](https://esdoc.org/manual/tags.html) to document the function:

```js
/**
 * Reorder the elements of an array using their indices.
 * @param {Array<T>} list - An array of elements to be reorder.
 * @param {number|Array<number>} originalIndices - Index or indices of the elements to be reordered. If multiple indices are provided the order they are specified in will be preserved in the resulting array.
 * @param {number} targetIndex - The index the reordered items should appear at in the resulting array.
 * @return {Array<T>} A new array with the items from the original array reordered.
 * @example
 * reorder(["a", "b", "c", "d"], [2, 0], 1); // -> ["b", "c", "a", "d"]
 */
```

I've noticed some tests cases that would fail with the current implementation:

```js
test("consecutive elements, move to end", () => {
  expect(reorder(["a", "b", "c"], [0, 1], 1)).toEqual(["c", "a", "b"]);
});

test("consecutive elements reordered, move to end", () => {
  expect(reorder(["a", "b", "c"], [1, 0], 1)).toEqual(["c", "b", "a"]);
});
```

I have quite worked out where/why that's happening, but I think there's some simplifying that can be done to that function. Conceptually, it needs to do 3 things:

1.  Get the arguments into a known shape — make the `originalIndices` argument an array of indices even when only a single index was supplied.
2.  Split `list` into two separate lists — the items being moved (in the order they need to be in) and the items not being moved (in the order they were originally in).
3.  Inject the items being moved into the items not being moved at the target index specified.

Unfortunately, JS doesn't give us a nice easy way to do step 2, so we'll have to do two passes over the list to separate it into the parts we want (this is achievable in utilities libraries such as Ramda using [groupBy](https://ramdajs.com/docs/#groupBy), but it's probably a bit heavy handed to introduce a whole dependency for one fairly minor case). We'll also need to do some funkiness to inject the items back into the array because the `splice` method takes the items to inject as individual arguments instead of an array. Here's the gist of how we can do it (I haven't actually tested this code and I expect there are some mistakes in it, but it should give you a good starting point):

```js
// reorder(["a", "b", "c", "d"], [2, 0], 1);
const originalIndicesAsArray = Array.isArray(originalIndices)
  ? originalIndices
  : [originalIndices];
const itemsBeingMoved = originalIndicesAsArray.map(i => list[i]);
const itemsNotMoving = list.filter(
  (_, i) => !originalIndicesAsArray.includes(i)
);
return Array.prototype.splice.apply(itemsNotMoving, [
  targetIndex,
  0,
  ...itemsBeingMoved
]);
```

Also, I see you are ignoring duplicates in your implementation. You're right, it's a case we need to handle, but I think it's probably better to just throw an error instead of trying to remove the duplicates. It's not really an intended use case and the exact results would be hard to predict (a user wouldn't know if it uses the position of the first or last duplicate index for the item's new position) and it makes our life a lot easier to explicitly not handle the case instead of adding lots of extra code to cover it off. We can just do something like this (and similar things for cases where there are indices that are out of bounds, although they'll probably throw errors anyway so it won't really add much value handling them ourselves):

```js
if (new Set(originalIndicesAsArray).size !== originalIndicesAsArray.length)
  throw TypeError("Duplicate indices are not allowed in originalIndices");
```

FYI — You can test that the function correctly throws an error with [toThrow](https://jestjs.io/docs/en/expect#tothrowerror).

Mxchaeltrxn commented:
Still gotta add the tests that check if the function throws correctly so this isn't done yet

haydn commented on src/reorder.js:
```diff
@@ -1,9 +1,30 @@
+  );
+
+  const reorderedList = itemsNotMoving;
+  reorderedList.splice(targetIndex, 0, ...itemsBeingMoved);
```
Nice!
